packages/macros/src/with_components.rs

```diff
@@ -0,0 +1,857 @@
+    let item_struct_children = item_struct_mnode.children.as_mut().unwrap();
+    let components_rnode =
+        ComponentsGenerationData(components_info).generate_for_storage_struct(db);
+
+    // Insert the components at the beginning of the struct body.
+    item_struct_children.insert(ast::ItemStruct::INDEX_LBRACE + 1, components_rnode);
+}
+
+/// Modifies the event enum to add the component events.
+fn process_event_enum(
+    item_enum: &mut RewriteNode,
+    db: &dyn SyntaxGroup,
+    components_info: &Vec<ComponentInfo>,
+) {
+    let item_enum_mnode = item_enum.modify(db);
+    let item_enum_children = item_enum_mnode.children.as_mut().unwrap();
+    let components_rnode = ComponentsGenerationData(components_info).generate_for_event_enum(db);
+
+    // Insert the components at the beginning of the enum body.
+    item_enum_children.insert(ast::ItemEnum::INDEX_LBRACE + 1, components_rnode);
+}
+
+fn add_event_enum(
+    body_rnode: &mut RewriteNode,
+    db: &dyn SyntaxGroup,
+    components_info: &Vec<ComponentInfo>,
+) {
+    let body_mnode = body_rnode.modify(db);
+    let event_enum_rnode = ComponentsGenerationData(components_info).generate_event_enum(db);
+
+    // It is safe to unwrap here because we know that the node has at least the storage struct children
+    body_mnode
+        .children
+        .as_mut()
+        .unwrap()
+        .insert(ast::ModuleBody::INDEX_RBRACE, event_enum_rnode);
+}
+
+/// Modifies the body node to add the use clauses and the `component!` macros to the module.
+fn add_use_clauses_and_macros(
+    body_rnode: &mut RewriteNode,
+    db: &dyn SyntaxGroup,
+    components_info: &Vec<ComponentInfo>,
+) {
+    let body_mnode = body_rnode.modify(db);
+    let components_rnode = ComponentsGenerationData(components_info).generate_for_module(db);
+
+    // It is safe to unwrap here because we know that the node has at least the storage struct children
+    body_mnode
+        .children
+        .as_mut()
+        .unwrap()
+        .insert(ast::ModuleBody::INDEX_RBRACE, components_rnode);
+}
+
+/// Information about a component.
+///
+/// # Members
+///
+/// * `name` - The name of the component (e.g. `ERC20Component`)
+/// * `path` - The path from where the component is imported (e.g. `openzeppelin_token::erc20::ERC20Component`)
+/// * `storage` - The path to reference the component in storage (e.g. `erc20`)
+/// * `event` - The path to reference the component events (e.g. `ERC20Event`)
+/// * `has_initializer` - Whether the component requires an initializer (e.g. `true`)
+/// * `internal_impls` - The internal implementations of the component to be added to
+///   the module by default (e.g. `["InternalImpl1", "InternalImpl2"]`)
+#[derive(Debug, Clone)]
+pub struct ComponentInfo {
+    pub name: String,
+    pub path: String,
+    pub storage: String,
+    pub event: String,
+    pub has_initializer: bool,
+    pub has_immutable_config: bool,
+    pub internal_impls: Vec<String>,
+}
```

Comment:
It looks like we don't need owned string in the code, so we could use `&str` type instead of `String` to make code cleaner and more performant. Especially the `get_component_info` will benefit from it if we can get rid of all the `to_string()` calls

```suggestion
pub struct ComponentInfo {
    pub name: &str,
    pub path: &str,
    pub storage: &str,
    pub event: &str,
    pub has_initializer: bool,
    pub has_immutable_config: bool,
    pub internal_impls: Vec<&str>,
}
```